### PR TITLE
feat: add a BillingConnector to send funds to a Billing contract on L2

### PR DESCRIPTION
**Pull request by pcarranzav**

### CODE REVIEW

File: contracts/Billing.sol [hunk 1]
```diff
@@ -59,9 +66,11 @@ contract Billing is IBilling, Governed {
     constructor(
         address _gateway,
```
Comment (abarmat) [inline contracts/Billing.sol:67]: This is unrelated to the scope of adding the token gateway (bridge) but I'd use the opportunity to rename `gateway` to `agent` or something different to avoid mistakes, the role is really a funds collection agent. In addition I think I'd make it so we can configure multiple agents.

File: contracts/Billing.sol [hunk 2]
```diff
@@ -1,18 +1,17 @@
 // SPDX-License-Identifier: MIT
 
-pragma solidity ^0.8.4;
+pragma solidity ^0.8.16;
```
Comment (abarmat) [inline contracts/Billing.sol:3]: Nice to use the latest solidity version

File: contracts/Billing.sol [hunk 3]
```diff
@@ -110,6 +121,17 @@ contract Billing is IBilling, Governed {
...
     }
 
+    /**
+     * @dev Add tokens into the billing contract for any user, from L1
+     * This can only be called from L2GraphTokenGateway.finalizeInboundTransfer.
+     * @param _user  Address that tokens are being added to
+     * @param _amount  Amount of tokens to add
+     */
+    function addFromL1(address _user, uint256 _amount) external override onlyL2TokenGateway {
+        userBalances[_user] = userBalances[_user] + _amount;
```
Comment (abarmat) [inline contracts/Billing.sol:131]: We could maybe leave this in `_add()`:
```
        userBalances[_user] = userBalances[_user] + _amount;
        emit TokensAdded(_user, _amount);
```
and have a `pullAndAdd()` with:
```
        require(_amount != 0, "Must add more than 0");
        require(_user != address(0), "user != 0");
        require(graphToken.transferFrom(_from, address(this), _amount), "Add transfer failed");
        _add(_user, _amount);
```

File: contracts/BillingConnector.sol [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        emit TokensSentToL2(msg.sender, _to, _amount);
+    }
+
+    /**
+     * @dev Allows the Governor to rescue any ERC20 tokens sent to this contract by accident
+     * @param _to  Destination address to send the tokens
+     * @param _token  Token address of the token that was accidentally sent to the contract
+     * @param _amount  Amount of tokens to pull
+     */
+    function rescueTokens(
```
Comment (abarmat) [inline contracts/BillingConnector.sol:124]: Maybe we can abstract this into a Rescatable class

File: contracts/BillingConnector.sol [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+     */
+    function setL1TokenGateway(address _l1TokenGateway) external override onlyGovernor {
+        _setL1TokenGateway(_l1TokenGateway);
+    }
+
+    /**
+     * @dev Sets the L2 Billing address
+     * @param _l2Billing New address for the L2 Billing contract
+     */
+    function setL2Billing(address _l2Billing) external override onlyGovernor {
```
Comment (abarmat) [inline contracts/BillingConnector.sol:73]: This is good as long as `onlyGovernor` is always the Council to avoid anyone to circumvent the gateway allowlist

File: contracts/BillingConnector.sol [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    function addToL2(
+        address _to,
+        uint256 _amount,
+        uint256 _maxGas,
+        uint256 _gasPriceBid,
+        uint256 _maxSubmissionCost
+    ) external payable override {
+        require(_amount != 0, "Must add more than 0");
+        require(_to != address(0), "destination != 0");
+        require(graphToken.transferFrom(msg.sender, address(this), _amount), "Add transfer failed");
```
Comment (abarmat) [inline contracts/BillingConnector.sol:95]: This might be a good opportunity to use the Permit feature from the GRT. We could add a `addToL2WithSig` so we can get the user to do everything with no token approval transaction, sign message, then send final transaction.

**Review by abarmat — COMMENTED** [review]
This is looking very good! I left a few comments

**Comment by abarmat** [comment]
@IainM22 @nenadjaja @brandonleekramer you can see a draft **BillingConnector** to bridge funds from L1 to an L2 billing

File: contracts/Billing.sol (hunk 1 above)
Comment (pcarranzav) [reply contracts/Billing.sol:67]: Good idea. How about `collector`? (and using a mapping of authorized collectors)

**Comment by pcarranzav** [comment]
Thanks for the comments @abarmat, I've implemented your suggestions and fixed the existing tests. I still need to add new tests for the new contract and features.

File: contracts/Billing.sol [hunk 7]
```diff
@@ -41,54 +44,59 @@ contract Billing is IBilling, Governed {
...
      */
-    function _setGateway(address _newGateway) internal {
-        require(_newGateway != address(0), "Gateway cannot be 0");
-        gateway = _newGateway;
-        emit GatewayUpdated(gateway);
+    function setCollector(address _collector, bool _enabled) external override onlyGovernor {
+        _setCollector(_collector, _enabled);
+    }
+
+    function setL2TokenGateway(address _l2TokenGateway) external onlyGovernor {
```
Comment (tmigone) [inline contracts/Billing.sol:98]:
```suggestion
    function setL2TokenGateway(address _l2TokenGateway) external override onlyGovernor {
```

File: contracts/IBilling.sol [hunk 8]
```diff
@@ -1,13 +1,14 @@
...
 interface IBilling {
     /**
-     * @dev Set the new gateway address
-     * @param _newGateway  New gateway address
+     * @dev Set or unset an address as an allowed Collector
+     * @param _collector  Collector address
+     * @param _enabled True to set the _collector address as a Collector, false to remove it
      */
-    function setGateway(address _newGateway) external; // onlyGateway or onlyGovernor, or something
+    function setCollector(address _collector, bool _enabled) external; // onlyGovernor
```
Comment (tmigone) [inline contracts/IBilling.sol:12]:
```suggestion

    /**
     * @dev Sets the L2 token gateway address
     * @param _l2TokenGateway New address for the L2 token gateway
     */
    function setL2TokenGateway(address _l2TokenGateway) external;

```

File: contracts/Billing.sol [hunk 9]
```diff
@@ -41,54 +44,59 @@ contract Billing is IBilling, Governed {
...
+     * @param _enabled True to set the _collector address as a Collector, false to remove it
      */
-    function _setGateway(address _newGateway) internal {
-        require(_newGateway != address(0), "Gateway cannot be 0");
-        gateway = _newGateway;
-        emit GatewayUpdated(gateway);
+    function setCollector(address _collector, bool _enabled) external override onlyGovernor {
+        _setCollector(_collector, _enabled);
+    }
+
```
Comment (tmigone) [inline contracts/Billing.sol:97]:
```suggestion
    
    /**
     * @dev Sets the L2 token gateway address
     * @param _l2TokenGateway New address for the L2 token gateway
     */
```

File: contracts/BillingConnector.sol [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+     */
+    event L1TokenGatewayUpdated(address l1TokenGateway);
+    /**
+     * @dev Address of the L2 Billing contract was updated
+     */
+    event L2BillingUpdated(address l2Billing);
+    /**
+     * @dev Tokens sent to the Billing contract on L2
+     */
+    event TokensSentToL2(address _from, address _to, uint256 _amount);
```
Comment (tmigone) [inline contracts/BillingConnector.sol:39]:
```suggestion
    event TokensSentToL2(address indexed _from, address indexed _to, uint256 _amount);
```

Maybe we want to index the addresses?

**Review by tmigone — COMMENTED** [review]
🚀 left a few minor suggestions

File: contracts/BillingConnector.sol [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+     */
+    event ArbitrumInboxUpdated(address inbox);
+    /**
+     * @dev Tokens sent to the Billing contract on L2
+     */
+    event TokensSentToL2(address indexed _from, address indexed _to, uint256 _amount);
+    /**
+     * @dev Request sent to the Billing contract on L2 to remove tokens from the balance
+     */
+    event TokensRemovedOnL2(address indexed _from, address indexed _to, uint256 _amount);
```
Comment (tmigone) [inline contracts/BillingConnector.sol:50]: I wonder if we should rename this event to something else given that as described below there is no guarantee that the tx on L2 will succeed. Maybe `RequestTokensRemovedOnL2` or something similar

**Review by tmigone — APPROVED** [review]

File: contracts/BillingConnector.sol (hunk 11 above)
Comment (pcarranzav) [reply contracts/BillingConnector.sol]: Good point, replaced for `RemovalRequestSentToL2`

**Review by tmigone — APPROVED** [review]
👍
